### PR TITLE
Event listener refactor

### PR DESCRIPTION
**Pull request by bitinn**
Address issues raised in #12, especially strong jQuery dependency.

Box team might have bigger fish to fry, so open a pull request here to signal someone is working on it.

----

Now it works, I end up doing following changes:

1. Abstract event system into `addListener` and `removeListener`, still prefer jQuery over native `add/removeEventListener`, but allow seamless fallback.

2. Update karma config and use `mocha` reporter instead of `progress`, this is good because `progress` does not show which test fails when you do `sandbox.verifyAndRestore` in `afterEach`; though generally one should avoid assert (`sandbox.verify` is implied assert) in `afterEach` because it cause mocha to abort all subsequent tests.

3. Update test cases so that it no longer depends on jQuery in global space, by `jQuery.noConflict` at the start. Note that we still test for jQuery global, see tests for `on[event]()`

4. Update travis to include node v0.12 and latest iojs

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @bitinn, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

**Comment by bitinn** [comment]
CLA signed

**Comment by boxcla** [comment]
Verified that @bitinn has just signed the CLA. Thanks, and we look forward to your contribution.

**Comment by bitinn** [comment]
@nzakas I think it's ready now, let me know if you find anything weird.

File: .travis.yml [hunk 1]
```diff
@@ -1,4 +1,6 @@
 language: node_js
 node_js:
     - "0.10"
-
+    - "0.12"
```
Comment (nzakas) [inline .travis.yml]: These changes are unnecessary. Since T3 runs in the browser, there's no benefit to running the tests in multiple Node environments.

File: config/karma-conf.js [hunk 2]
```diff
@@ -67,7 +67,7 @@ module.exports = function(config) {
 		// test results reporter to use
 		// possible values: 'dots', 'progress'
 		// available reporters: https://npmjs.org/browse/keyword/karma-reporter
-		reporters: ['progress', 'coverage', 'threshold'],
+		reporters: ['mocha', 'coverage', 'threshold'],
```
Comment (nzakas) [inline config/karma-conf.js]: Why was this changed?

File: tests/application-test.js [hunk 3]
```diff
@@ -7,11 +7,19 @@ describe('Box.Application', function() {
...
 	var sandbox = sinon.sandbox.create();
 
 	var testModule,
 		testModule2,
-		nestedModule;
+		nestedModule,
+		customEvent;
+
+	var dispatch = function(selector) {
+		customEvent = document.createEvent('MouseEvent');
```
Comment (nzakas) [inline tests/application-test.js]: This will fail in IE8.

File: .travis.yml (hunk 1 above)
Comment (bitinn) [inline .travis.yml]: I believe it's important to make sure the development environment is compatible to latest version of node.js and io.js

File: config/karma-conf.js (hunk 2 above)
Comment (bitinn) [inline config/karma-conf.js]: To quote my item 2:

Update karma config and use mocha reporter instead of progress, this is good because progress does not show which test fails when you do sandbox.verifyAndRestore in afterEach; though generally one should avoid assert (sandbox.verify is implied assert) in afterEach because it cause mocha to abort all subsequent tests.

**Comment by nzakas** [comment]
Thanks for using the time to do this. I made some initial comments, but I'll discuss with the team as well.

Generally, we like to have just one commit per pull request and we like to keep unrelated changes separate. If you'd like to propose changing the Karma reporter, then please file a separate issue for that - it should not be bundled with this change.

File: tests/application-test.js (hunk 3 above)
Comment (bitinn) [inline tests/application-test.js]: ok, but to test `addEventListener` fallback, you need IE9+

**Comment by bitinn** [comment]
@nzakas sure, I can rebase it, just need to make sure you are ok with the pull request in general.

please see my reply, I think the change make sense, for an example of `mocha` report, see http://mochajs.org/#reporters, it's the default `spec` reporter.

also see https://www.npmjs.com/package/karma-mocha-reporter for screenshot.

File: tests/application-test.js (hunk 3 above)
Comment (bitinn) [inline tests/application-test.js]: this is a dilemma really:

1. if you want to the test to run in IE8, then jQuery is needed.

2. but if you want to test `addEventListener`, then jQuery must not in the global, at least with current fallback design.

**Comment by nzakas** [comment]
As I mentioned, the change to the karma reporter should be requested in a separate issue. I've not seen any trouble using the progress reporter when tests fail, so if you are having trouble, it needs to be addressed separately.

**Comment by bitinn** [comment]
No problem, I will take it out soon, and open a new ticket for it.

**Comment by bitinn** [comment]
I have removed unrelated changes, though still need your decision on whether it's necessary to run tests in IE8.

AFAIK: `jQuery.trigger` only trigger `jQuery.on` handlers, and `dispatchEvent` only trigger `addEventListener` handlers, so 2 sets of tests needed.

What I currently do is to mainly test with `dispatchEvent` and `addEventListener`, with 1 additional test to make sure jQuery works (we know it works).

**Comment by bitinn** [comment]
A suggestion: we check for `querySelectorAll` support before `var $ = jQuery.noConflict(true)`, so for IE8 we assume global $ exists, and make test cases use jQuery.trigger instead of dispatchEvent.

If you are ok with that, I will go ahead and update the test cases.

**Comment by nzakas** [comment]
Yes, we must run tests in IE8. Still wrangling the team to review the code.

**Comment by nzakas** [comment]
The team reviewed this change proposal and the general feeling is that we don't want the core to be doing jQuery detection. We've started https://github.com/box/t3js/issues/22 to continue discussing.
